docs/source/yosys_internals/formats/cell_library.rst

```diff
@@ -619,6 +619,52 @@ Finite state machines
 
 Add a brief description of the ``$fsm`` cell type.
 
+Coarse arithmetics
+~~~~~~~~~~~~~~~~~~~~~
+
+The ``$macc`` cell type represents a multiply and accumulate block, for summing any number of negated and unnegated signals and arithmetic products of pairs of signals. Cell port A concatenates pairs of signals to be multiplied together. When the second signal in a pair is zero length, a constant 1 is used instead as the second factor. Cell port B concatenates 1-bit-wide signals to also be summed, such as "carry in" in adders. 
+
+The cell's ``CONFIG`` parameter determines the layout of cell port ``A``.
+In the terms used for this cell, there's mixed meanings for the term "port". To disambiguate:
+A cell port is for example the A input (it is constructed in C++ as ``cell->setPort(ID::A, ...))``
+Multiplier ports are pairs of multiplier inputs ("factors").
+If the second signal in such a pair is zero length, no multiplication is necessary, and the first signal is just added to the sum.
+
+In this pseudocode, ``u(foo)`` means an unsigned int that's foo bits long.
+The CONFIG parameter carries the following information:
+.. code-block::
+	:force:
+	struct CONFIG {
+		u4 num_bits;
+		struct port_field {
+			bool is_signed;
+			bool is_subtract;
+			u(num_bits) factor1_len;
+			u(num_bits) factor2_len;
+		}[num_ports];
+	};
+
+The A cell port carries the following information:
+.. code-block::
+	:force:
+	struct A {
+		u(CONFIG.port_field[0].factor1_len) port0factor1;
+		u(CONFIG.port_field[0].factor2_len) port0factor2;
+		u(CONFIG.port_field[1].factor1_len) port1factor1;
+		u(CONFIG.port_field[1].factor2_len) port1factor2;
+		...
+	};
+
+No factor1 may have a zero length.
+A factor2 having a zero length implies factor2 is replaced with a constant 1.
+
+Additionally, B is an array of 1-bit-wide unsigned integers to also be summed up.
+Finally, we have:
+.. code-block::
+	:force:
```

Comment:
Why are all the code-blocks using the `:force:` option?  Also `.. code-block::` should have a new line before it to prevent errors.

Reply:
I removed force in the fixup commit since it didn't change anything. It's used elsewhere in the file and I haven't played with that

Reply:
Oh huh, I didn't actually see that the rest of them use it too... I think those came from the auto conversion and I haven't looked at this page since...